lib/schemes/refresh.js

```diff
@@ -0,0 +1,329 @@
+    return this.$auth.requestWith(
+      this.name,
+      endpoint,
+      this.options.endpoints.refresh
+    ).then(response => {
+      this._updateTokens(response)
+    }).catch(() => {
+      // TODO: Unhandled error
+      this._logoutLocally()
+    })
+  }
+
+  _scheduleTokenRefresh () {
+    // If auto refresh is disabled, bail
+    if (this.options.disableAutoRefresh) return
+
+    let intervalDuration = (this._getExpiration() - Date.now()) * 0.75
+    if (intervalDuration < 1000) {
+      // in case you misconfigured refreshing this will save your auth-server from a self-induced DDoS-Attack
+      intervalDuration = 1000
+    }
+
+    this.refreshInterval = setInterval(() => {
+      this._tokenRefresh()
+    }, intervalDuration)
+  }
+
+  _initializeErrorInterceptor () {
+    this.$auth.ctx.app.$axios.onError(error => {
+      // If error has config and response and error status is 401
+      if (error.config && error.response && error.response.status === 401) {
+        const token = this.$auth.getToken(this.name)
+        const headerToken = error.config.headers[this.options.tokenName]
+
+        // If current token is equal to header token
+        if (token && headerToken && headerToken === token) {
+          // Add the failed request config to failedRequestQueue
+          if (!error.config.__isRetryRequest) {
+            error.config.__isRetryRequest = true
+            this.failedRequestQueue.push(error.config)
+          }
+
+          // If is not refreshing the token
+          if (!this.isRefreshing) {
+            // Refresh it
+            return this._tokenRefresh().then(() => {
+              this._syncExpiration()
+              this._scheduleTokenRefresh()
+            })
+          } else {
+            // Or resolve Promise
+            return Promise.resolve()
+          }
+        } else if (token && headerToken && headerToken !== token && !error.config.__isRetryRequest) {
+          // Or just retry the request if token is not equal to header token
+          error.config.__isRetryRequest = true
+          error.config.headers[this.options.tokenName] = token
+          return this.$auth.ctx.app.$axios.request(error.config)
+        }
+      }
+
+      // Otherwise, reject Promise
+      return Promise.reject(error)
+    })
+  }
+
+  async mounted () {
+    if (this.options.tokenRequired) {
+      const token = this.$auth.syncToken(this.name)
+      this._setToken(token)
+      this._syncRefreshToken()
+
+      const expiration = this._syncExpiration()
+      if (Date.now() >= expiration) {
+        await this._logoutLocally()
+      }
+
+      this.$auth.syncClientId(this.name)
+    }
+
+    return this.$auth.fetchUserOnce().then((response) => {
+      // Initialize axios error interceptor
+      this._initializeErrorInterceptor()
+
+      // Only refresh token if user is logged in and is client side
+      if (process.client && this.$auth.loggedIn) {
+        setTimeout(() => {
```

Comment:
Why do we wait for a second before refreshing? Maybe we should include a comment on this might-be-race-condition-quirk ? 🙂

Reply:
This is to avoid conflict with page request. Without that the refresh would be sent at the same time of the first page request, causing an error in secure pages.
I mean the request of the page itself.

Reply:
I don't quite understand why a token request will conflict with the current page request. This happens client side, after the user data has been fetched.
The auth module is initializing, and the page render has not begun yet. If we remove the 1 second timeout before refreshing, the tokens will potentially be replaced by new tokens before the page renders. How exactly can this cause an error in secured pages?

Reply:
Ok, I'll test it again. 
But I remember that I needed to use this to fix something.

Reply:
Great! To a minimum I think we should include a comment to explain why this is a race condition :)